### PR TITLE
Fix potential delay in sending out requests from the rust SDK

### PR DESCRIPTION
**Pull request by richvdh**
There was a potential race which could cause us to be very slow to send out pending HTTP requests, particularly when handling a user verification. Add some resilience to make sure we handle it correctly.

~~Based on #3716~~

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

**Comment by andybalaam** [comment]
Updated from develop to make the diff clearer, since the based-on PR is merged. (Hope that's ok)

File: src/rust-crypto/rust-crypto.ts [hunk 1]
```diff
@@ -1323,24 +1335,43 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+            //
+            // There is potential for a race whereby the item is added *after* `OlmMachine.outgoingRequests` checks
+            // the queue, but *before* it returns. In such a case, the item could sit there unnoticed for some time.
+            //
+            // In order to circumvent the race, we set a flag which tells the loop to go round once again even if the
+            // queue appears to be empty.
+            this.outgoingRequestLoopOneMoreLoop = true;
             return;
         }
+        this.outgoingRequestLoopInner().catch((e) => {
```
Comment (andybalaam) [inline src/rust-crypto/rust-crypto.ts:1351]: Worth a comment saying we are deliberately not awaiting this here?

File: src/rust-crypto/rust-crypto.ts [hunk 2]
```diff
@@ -1323,24 +1335,43 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+
+    private async outgoingRequestLoopInner(): Promise<void> {
+        /* istanbul ignore if */
+        if (this.outgoingRequestLoopRunning) {
+            throw new Error("Cannot run two outgoing request loops");
+        }
         this.outgoingRequestLoopRunning = true;
         try {
             while (!this.stopped) {
+                this.outgoingRequestLoopOneMoreLoop = false;
```
Comment (weeman1337) [inline src/rust-crypto/rust-crypto.ts:1364]: Can you add another short comment here? Because if you oversee the `await` one may think that `this.outgoingRequestLoopOneMoreLoop` is always false in the `if` below.

**Review by andybalaam — APPROVED** [review]
Looks sensible and as described. Any way to test this?

**Review by weeman1337 — APPROVED** [review]
Looks good to me. I've also tested device verification on the deployment. Works.

I suppose that writing tests for this edge case is not easily doable?
